clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -163,6 +175,17 @@ public void run() {
      */
     void run(long now) {
         Cluster cluster = metadata.fetch();
+
+        if (transactionState.isIdempotenceEnabled()) {
+            while (!transactionState.pidIsSet() && (time.milliseconds() - now) < maxBlockTimeMs) {
```

Comment:
I'm not sure about the use of `maxBlockTimeMs` here. My understanding is that this is used to control how long the user will wait in the worst case for `send`, but this is happening in the background. Maybe we should keep trying indefinitely until we've received a PID?